servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/JavaNetSoTimeoutHttpConnectionFilter.java

```diff
@@ -167,24 +174,99 @@ public Single<StreamingHttpResponse> request(final StreamingHttpRequest request)
-                                            .onErrorMap(TimeoutException.class, t -> newStacklessSocketTimeoutException(
-                                                    "Read timed out after " + timeout.toMillis() +
-                                                            "ms waiting for response meta-data")
-                                                    .initCause(t))))
-                            .map(response -> response.transformMessageBody(p -> p.timeout(timeout, timeoutExecutor)
-                                    .onErrorMap(TimeoutException.class, t -> newStacklessSocketTimeoutException(
-                                            "Read timed out after " + timeout.toMillis() +
-                                                    "ms waiting for the next response payload body chunk")
-                                            .initCause(t))))
+                            .<StreamingHttpResponse>liftSync(subscriber ->
+                                    new RequestTimeoutSubscriber(subscriber,
+                                            SourceAdapters.fromSource(requestProcessor), timeout, timeoutExecutor))
                             .shareContextOnSubscribe();
                 });
             }
         };
     }
 
+    // package private for testing purposes
+    static final class RequestTimeoutSubscriber implements Subscriber<StreamingHttpResponse> {
+
+        private static final AtomicIntegerFieldUpdater<RequestTimeoutSubscriber> onceUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(RequestTimeoutSubscriber.class, "once");
+
+        private final DelayedCancellable requestCancellable = new DelayedCancellable();
+        private final Cancellable timeoutCancellable;
+        private final Subscriber<? super StreamingHttpResponse> delegate;
+
+        private final Duration timeout;
+        private final Executor timeoutExecutor;
+        @SuppressWarnings("unused")
+        private volatile int once;
+
+        RequestTimeoutSubscriber(Subscriber<? super StreamingHttpResponse> delegate, Completable requestComplete,
+                                 Duration timeout, Executor timeoutExecutor) {
+            this.delegate = delegate;
+            this.timeout = timeout;
+            this.timeoutExecutor = timeoutExecutor;
+            timeoutCancellable = requestComplete.concat(Completable.never()
+                    .timeout(timeout, timeoutExecutor)).beforeOnError(this::handleInterruptions).subscribe();
+        }
+
+        @Override
+        public void onSubscribe(Cancellable cancellable) {
+            requestCancellable.delayedCancellable(cancellable);
+            try {
+                delegate.onSubscribe(() -> {
+                    once();
+                    timeoutCancellable.cancel();
+                    requestCancellable.cancel();
+                });
+            } catch (Throwable cause) {
```

Comment:
Based on the chat discussion: no need to try-catch `onSubscribe` inside operators because the exception will go up to the original source and that source is responsible for try-catching, cleaning up its own state (no need to cancel `cancellable`), and propagating `onError`, which will result in `timeoutCancellable.cancel()` from there.

Reply:
After reviewing all other operators as part of #3046, I'm agreed that `requestCancellable.delayedCancellable(cancellable)` should be after `onSubscribe`. Thanks for bringing it back.

> I don't think that's quite right because we've already initiated the timer so we need to try / catch to make sure we've cleaned that up. 

As Scott described in chat, the unexpected exception from `onSubscribe` always goes up to the original source, and it will propagate it down via `onError`, which will call `timeoutCancellable.cancel()` if it arrives before the timer expires. Why do you think it's not enough? `TimeoutSingle` relies on the same principles and it starts the timer in the constructor.

Reply:
I ended up moving the initialization of the timer to after the `onSubscribe(..)` call to avoid the potential race between the timer firing and the `subscriber.onSubscribe(..)` call which would violate the RS spec which sidesteps the problem of needing to cleanup the resource.

Reply:
That's a good catch! Consider adding a comment to clarify that and make sure nobody accidentally moves it. Now I see that `TimeoutSingle` uses more complex state machine to prevent that, we don't need to repeat that here.

Reply:
Done.